test/test_ekfac.py

```diff
@@ -0,0 +1,1094 @@
+        separate_weight_and_bias=separate_weight_and_bias,
+    )
+    ekfac_mat = ekfac.to_scipy() @ eye(ekfac.shape[1])
+
+    atol = {"sum": 1e-3, "mean": 1e-3}[loss_func.reduction]
+    rtol = {"sum": 3e-2, "mean": 3e-2}[loss_func.reduction]
+
+    report_nonclose(ggn, ekfac_mat, rtol=rtol, atol=atol)
+
+
+@mark.parametrize(
+    "separate_weight_and_bias", [True, False], ids=["separate_bias", "joint_bias"]
+)
+@mark.parametrize(
+    "exclude", [None, "weight", "bias"], ids=["all", "no_weights", "no_biases"]
+)
+@mark.parametrize("shuffle", [False, True], ids=["", "shuffled"])
+def test_ekfac_ef_one_datum(
+    kfac_exact_one_datum_case: Tuple[
+        Module,
+        Union[BCEWithLogitsLoss, CrossEntropyLoss],
+        List[Parameter],
+        Iterable[Tuple[Tensor, Tensor]],
+    ],
+    separate_weight_and_bias: bool,
+    exclude: str,
+    shuffle: bool,
+):
+    model, loss_func, params, data, batch_size_fn = kfac_exact_one_datum_case
+    params = maybe_exclude_or_shuffle_parameters(params, model, exclude, shuffle)
+
+    ef = block_diagonal(
+        EFLinearOperator,
+        model,
+        loss_func,
+        params,
+        data,
+        batch_size_fn=batch_size_fn,
+        separate_weight_and_bias=separate_weight_and_bias,
+    )
+
+    ekfac = EKFACLinearOperator(
+        model,
+        loss_func,
+        params,
+        data,
+        batch_size_fn=batch_size_fn,
+        fisher_type=FisherType.EMPIRICAL,
+        separate_weight_and_bias=separate_weight_and_bias,
+    )
+    ekfac_mat = ekfac.to_scipy() @ eye(ekfac.shape[1])
+
+    report_nonclose(ef, ekfac_mat, atol=1e-7)
+
+
+@mark.parametrize("dev", DEVICES, ids=DEVICES_IDS)
+def test_ekfac_inplace_activations(dev: device):
+    """Test that EKFAC works if the network has in-place activations.
+
+    We use a test case with a single datum as EKFAC becomes exact as the number of
+    MC samples increases.
+
+    Args:
+        dev: The device to run the test on.
+    """
+    manual_seed(0)
+    model = Sequential(Linear(6, 3), ReLU(inplace=True), Linear(3, 2)).to(dev)
+    loss_func = MSELoss().to(dev)
+    batch_size = 1
+    data = [(rand(batch_size, 6), regression_targets((batch_size, 2)))]
+    params = list(model.parameters())
+
+    # 1) compare EKFAC and GGN
+    ggn = block_diagonal(GGNLinearOperator, model, loss_func, params, data)
+
+    ekfac = EKFACLinearOperator(model, loss_func, params, data, mc_samples=2_000)
+    ekfac_mat = ekfac.to_scipy() @ eye(ekfac.shape[1])
+
+    atol = {"sum": 5e-1, "mean": 2e-3}[loss_func.reduction]
+    rtol = {"sum": 2e-2, "mean": 2e-2}[loss_func.reduction]
+
+    report_nonclose(ggn, ekfac_mat, rtol=rtol, atol=atol)
+
+    # 2) Compare GGN (inplace=True) and GGN (inplace=False)
+    for mod in model.modules():
+        if hasattr(mod, "inplace"):
+            mod.inplace = False
+    ggn_no_inplace = block_diagonal(GGNLinearOperator, model, loss_func, params, data)
+
+    report_nonclose(ggn, ggn_no_inplace)
```

Comment:
This test should be refactored to share the same code as the equivalent test for KFAC.